infer/environment.cc

```diff
@@ -1281,7 +1282,7 @@ core::TypePtr Environment::processBinding(core::Context ctx, const cfg::CFG &inW
 
                 const core::TypeAndOrigins &ty = getAndFillTypeAndOrigin(ctx, c->value);
                 ENFORCE(c->cast != core::Names::uncheckedLet());
```

Comment:
Do you notice how here we assert that `uncheckedLet` doesn't make it this far? That's because we remove references to `uncheckedLet` in the `builder_walk` phase so that instead of 4 kinds of casts to consider in `infer`, there are only 3.

Can you change builder walk to effectively convert `T.bind`-style cast `Instruction`s to `T.cast`-style cast `Instruction`s? This way, we can be sure that future changes to `infer` never have to handle `T.bind` specially, and thus the behavior of `T.cast` and `T.bind` never diverge in ways we don't want them to.